domain/application/service/application.go

```diff
@@ -1208,3 +1208,26 @@ func (s *Service) GetApplicationsWithPendingCharmsFromUUIDs(ctx context.Context,
 	}
 	return s.st.GetApplicationsWithPendingCharmsFromUUIDs(ctx, uuids)
 }
+
+// ReserveCharmDownload reserves a charm download slot for the specified
+// application. If the charm is already being downloaded, the method will
+// return [applicationerrors.AlreadyDownloadingCharm]. The charm download
+// information is returned which includes the charm name, origin and the
+// digest.
+func (s *Service) ReserveCharmDownload(ctx context.Context, appID coreapplication.ID) (application.CharmDownloadInfo, error) {
+	if err := appID.Validate(); err != nil {
+		return application.CharmDownloadInfo{}, internalerrors.Errorf("application ID: %w", err)
+	}
+
+	return application.CharmDownloadInfo{}, errors.NotImplementedf("ReserveCharmDownload")
```

Comment:
Comming soon in a future PR.

Reply:
What does "reserve a slot" exactly means? I'm trying to think why we would need this usecase.